pkg/server/commands/check_command.go

```diff
@@ -27,25 +27,83 @@ const (
 	defaultMaxConcurrentReadsForCheck = math.MaxUint32
 )
 
+// CheckCommand interface wraps the Execute method for performing a check operation.
+type CheckCommand interface {
+	Execute(ctx context.Context) (*graph.ResolveCheckResponse, *graph.ResolveCheckRequestMetadata, error)
+}
+
 type CheckQuery struct {
+	params                     CheckCommandParams
 	logger                     logger.Logger
 	checkResolver              graph.CheckResolver
-	typesys                    *typesystem.TypeSystem
-	datastore                  storage.RelationshipTupleReader
 	sharedCheckResources       *shared.SharedDatastoreResources
 	cacheSettings              config.CacheSettings
 	maxConcurrentReads         uint32
 	shouldCacheIterators       bool
 	datastoreThrottleThreshold int
 	datastoreThrottleDuration  time.Duration
+	datastoreWithTupleCache    *storagewrappers.RequestStorageWrapper
 }
 
+var _ CheckCommand = (*CheckQuery)(nil)
+
 type CheckCommandParams struct {
 	StoreID          string
 	TupleKey         *openfgav1.CheckRequestTupleKey
 	ContextualTuples *openfgav1.ContextualTupleKeys
 	Context          *structpb.Struct
 	Consistency      openfgav1.ConsistencyPreference
+	Typesys          *typesystem.TypeSystem
+	Operation        string
+}
+
+// CheckCommandConfig server config required to run a CheckCommand.
+type CheckCommandConfig struct {
+	datastore     storage.RelationshipTupleReader
+	checkResolver graph.CheckResolver
+	options       []CheckQueryOption
+}
+
+// NewCheckCommandConfig creates a new CheckCommandConfig.
+func NewCheckCommandConfig(datastore storage.RelationshipTupleReader, checkResolver graph.CheckResolver, options ...CheckQueryOption) CheckCommandConfig {
+	return CheckCommandConfig{
+		datastore:     datastore,
+		checkResolver: checkResolver,
+		options:       options,
+	}
+}
+
+// CheckCommandServerConfig holds the full server configuration including shadow mode settings.
+type CheckCommandServerConfig struct {
+	checkCfg  CheckCommandConfig
+	shadowCfg ShadowCheckCommandConfig
+}
+
+type CheckCommandServerConfigOption func(*CheckCommandServerConfig)
+
+func WithShadowCheckCommandConfig(shadowCfg ShadowCheckCommandConfig) CheckCommandServerConfigOption {
+	return func(c *CheckCommandServerConfig) {
+		c.shadowCfg = shadowCfg
+	}
+}
+
+// NewCheckCommandServerConfig creates a new CheckCommandServerConfig.
+func NewCheckCommandServerConfig(main CheckCommandConfig, opts ...CheckCommandServerConfigOption) CheckCommandServerConfig {
```

Comment:
Why value rather than pointer like everywhere else? If its more performance/nicer to GC, make a subsequent PR were everything changes uniformly but introduce it as the rest of the things.